## Preparation
___

### Importing data
Importing necessary libraries and the dataset

In [1]:
import pandas as pd

In [2]:
from datetime import datetime

In [3]:
df = pd.read_csv('online_retail.csv', encoding='windows-1252')

### Exploring basic dataset information
Checking data types, dataset shape, null values. Previewing samples of data

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [5]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [6]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [7]:
df.sample(15)

InvoiceNo StockCode                       Description  Quantity  \
507205    579168    79190B          RETRO PLASTIC POLKA TRAY        30   
88922     543824     22748         POPPY'S PLAYHOUSE KITCHEN        60   
414990    572513     21401                BLUE PUDDING SPOON        12   
15580     537638    84913A             SOFT PINK ROSE TOWEL          2   
301693    563351     22681     FRENCH BLUE METAL DOOR SIGN 6        20   
217657    555929    82494L       WOODEN FRAME ANTIQUE WHITE         24   
425948    573340     21216  SET 3 RETROSPOT TEA,COFFEE,SUGAR         1   
155196    549977     20725           LUNCH BAG RED RETROSPOT         4   
47614     540418     21774   DECORATIVE CATS BATHROOM BOTTLE         4   
10984     537240     22798  ANTIQUE GLASS DRESSING TABLE POT         4   
329964    565862     23407    SET OF 2 TRAYS HOME SWEET HOME         2   
138869    548215     47566                     PARTY BUNTING         8   
503062    578858     22840        ROUND CAKE TIN VINTAGE RED         1   
292268    562545     22178   VICTORIAN GLASS HANGING T-LIGHT        96   
327094    565617     23119     PACK OF 6 LARGE FRUIT STRAWS         12   

           InvoiceDate  UnitPrice  CustomerID         Country  
507205  11/28/11 14:12       0.42     17806.0  United Kingdom  
88922     2/14/11 9:31       1.85     14631.0  United Kingdom  
414990  10/24/11 14:36       0.12     16728.0  United Kingdom  
15580    12/7/10 15:28       3.36         NaN  United Kingdom  
301693   8/15/11 14:17       1.25     16902.0  United Kingdom  
217657     6/8/11 8:12       2.55     16525.0  United Kingdom  
425948  10/30/11 11:55       4.95     14159.0  United Kingdom  
155196   4/13/11 14:54       1.65     14810.0  United Kingdom  
47614     1/7/11 11:04       1.28         NaN  United Kingdom  
10984    12/6/10 10:08       5.91         NaN  United Kingdom  
329964    9/7/11 14:56       9.95     15885.0  United Kingdom  
138869    3/30/11 9:19       4.95     15674.0  United Kingdom  
503062  11/27/11 11:22       7.95     18145.0  United Kingdom  
292268    8/5/11 15:46       1.06     12971.0  United Kingdom  
327094    9/5/11 15:30       1.25         NaN  United Kingdom

### Removing Duplicates

In [9]:
# Checking for duplicated rows
df.duplicated().value_counts()

False    536641
True       5268
Name: count, dtype: int64

In [10]:
# Removing duplicates
df = df.drop_duplicates()

___
### Resolving null values and data types:
___
⚠️ `CustomerID` is being treated as a numerical rather than categorical value

⚠️ `CustomerID` and `Description` fields have some null values. 

- Have to handle null values of `CustomerID`, otherwise won't be able to cast it into a string data type.
- Nulls in the `Description` column won't let me to filter data by regular expressions.
- Need to convert `InvoiceDate` column to a conventional format to make sure a predictable behaviour 
___

In [11]:
# Filling null values in the CustomerID column with zeros
df['CustomerID'] = df['CustomerID'].fillna(0)

In [12]:
# Filling null values in the Description column with zeros passed as string
df['Description'] = df['Description'].fillna('0')

In [14]:
# First, converting float to integer to get rid of zero decimal points from the customer ID
df['CustomerID'] = df['CustomerID'].astype("int64")

In [15]:
# Next, converting CustomerID from numeric to categorical value
df['CustomerID'] = df['CustomerID'].astype(str)

In [16]:
# Converting InvoiceDate column to a proper datetime format 
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m/%d/%y %H:%M')

In [17]:
# Removing :00 seconds trail for better readability
df['InvoiceDate'] = df['InvoiceDate'].apply(lambda t: t.strftime('%Y-%m-%d %H:%M'))

In [18]:
# Checking the result is correct
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    536641 non-null  object 
 1   StockCode    536641 non-null  object 
 2   Description  536641 non-null  object 
 3   Quantity     536641 non-null  int64  
 4   InvoiceDate  536641 non-null  object 
 5   UnitPrice    536641 non-null  float64
 6   CustomerID   536641 non-null  object 
 7   Country      536641 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 36.8+ MB


In [19]:
df[df.CustomerID == '0']

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                                0        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

             InvoiceDate  UnitPrice CustomerID         Country  
622     2010-12-01 11:52       0.00          0  United Kingdom  
1443    2010-12-01 14:32       2.51          0  United Kingdom  
1444    2010-12-01 14:32       2.51          0  United Kingdom  
1445    2010-12-01 14:32       0.85          0  United Kingdom  
1446    2010-12-01 14:32       1.66          0  United Kingdom  
...                  ...        ...        ...             ...  
541536  2011-12-09 10:26       4.13          0  United Kingdom  
541537  2011-12-09 10:26       4.13          0  United Kingdom  
541538  2011-12-09 10:26       4.96          0  United Kingdom  
541539  2011-12-09 10:26      10.79          0  United Kingdom  
541540  2011-12-09 10:26    1714.17          0  United Kingdom  

[135037 rows x 8 columns]

## Data Cleaning
___

In [20]:
# Generating descriptive statistics of numerical variables to check for potential issues
df.describe()

Quantity      UnitPrice
count  536641.000000  536641.000000
mean        9.620029       4.632656
std       219.130156      97.233118
min    -80995.000000  -11062.060000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        10.000000       4.130000
max     80995.000000   38970.000000

___
### There are some issues: 

- ⚠️ negative values for both `Quantity` and `UnitPrice` variables.
  *Order return? Typical practice is marking a returned product in the database by multiplying its quantity amount by (-1).*
- ⚠️ maximum and minumin values of `Quantity` variable are waaaay out of mean, and suspiciously equal.
  Wrong transaction that has been reversed?
- ⚠️ max and min of `UnitPrice` also look wrong and require exploring.
___

In [21]:
df[(df['Quantity'] == df['Quantity'].min()) | (df['Quantity'] == df['Quantity'].max())]

InvoiceNo StockCode                  Description  Quantity  \
540421    581483     23843  PAPER CRAFT , LITTLE BIRDIE     80995   
540422   C581484     23843  PAPER CRAFT , LITTLE BIRDIE    -80995   

             InvoiceDate  UnitPrice CustomerID         Country  
540421  2011-12-09 09:15       2.08      16446  United Kingdom  
540422  2011-12-09 09:27       2.08      16446  United Kingdom

In [22]:
# Removing both rows with clearly incorrect data
df = df.drop([540421, 540422])

In [23]:
# Checking statistics again
df.describe()

Quantity      UnitPrice
count  536639.000000  536639.000000
mean        9.620065       4.632665
std       153.521936      97.233299
min    -74215.000000  -11062.060000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        10.000000       4.130000
max     74215.000000   38970.000000

In [24]:
# There are another set of records with the same issue where max = |min| way too big. 
# Repeating previous steps:
df[(df['Quantity'] == df['Quantity'].min()) | (df['Quantity'] == df['Quantity'].max())]

InvoiceNo StockCode                     Description  Quantity  \
61619    541431     23166  MEDIUM CERAMIC TOP STORAGE JAR     74215   
61624   C541433     23166  MEDIUM CERAMIC TOP STORAGE JAR    -74215   

            InvoiceDate  UnitPrice CustomerID         Country  
61619  2011-01-18 10:01       1.04      12346  United Kingdom  
61624  2011-01-18 10:17       1.04      12346  United Kingdom

In [25]:
df = df.drop([61619, 61624])

In [26]:
df.describe()

Quantity      UnitPrice
count  536637.000000  536637.000000
mean        9.620101       4.632679
std        55.151517      97.233480
min     -9600.000000  -11062.060000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        10.000000       4.130000
max     12540.000000   38970.000000

In [27]:
df.sort_values(by='Quantity')

InvoiceNo StockCode                          Description  Quantity  \
225530    556691     23005         printing smudges/thrown away     -9600   
225529    556690     23005         printing smudges/thrown away     -9600   
4287     C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
225528    556687     23003         Printing smudges/thrown away     -9058   
115818    546152    72140F                           throw away     -5368   
...          ...       ...                                  ...       ...   
220843    556231    85123A                                    ?      4000   
206121    554868     22197                 SMALL POPCORN HOLDER      4300   
421632    573008     84077    WORLD WAR 2 GLIDERS ASSTD DESIGNS      4800   
74614     542504     37413                                    0      5568   
502122    578841     84826       ASSTD DESIGN 3D PAPER STICKERS     12540   

             InvoiceDate  UnitPrice CustomerID         Country  
225530  2011-06-14 10:37       0.00          0  United Kingdom  
225529  2011-06-14 10:37       0.00          0  United Kingdom  
4287    2010-12-02 14:23       0.03      15838  United Kingdom  
225528  2011-06-14 10:36       0.00          0  United Kingdom  
115818  2011-03-09 17:25       0.00          0  United Kingdom  
...                  ...        ...        ...             ...  
220843  2011-06-09 15:04       0.00          0  United Kingdom  
206121  2011-05-27 10:52       0.72      13135  United Kingdom  
421632  2011-10-27 12:26       0.21      12901  United Kingdom  
74614   2011-01-28 12:03       0.00          0  United Kingdom  
502122  2011-11-25 15:57       0.00      13256  United Kingdom  

[536637 rows x 8 columns]

In [28]:
df[df.Description.str.contains('throw', regex=False, case=True)]

InvoiceNo StockCode                    Description  Quantity  \
50849     540564     22617           mouldy, thrown away.     -2600   
82794     543257    84611B                    thrown away     -1430   
82796     543259    84612B                    thrown away      -162   
82797     543260    84613A                    thrown away      -230   
82799     543262    84614A                    thrown away      -390   
89118     543827    84406B  incorrectly made-thrown away.       -64   
115489    546126    35611B                    thrown away       -27   
115504    546129    35610B                    thrown away      -541   
115505    546130    35610A                    thrown away     -1277   
115818    546152    72140F                     throw away     -5368   
131398    547559     72759        thrown away-can't sell.      -524   
131399    547560     72732         thrown away-can't sell     -2472   
225528    556687     23003   Printing smudges/thrown away     -9058   
225529    556690     23005   printing smudges/thrown away     -9600   
225530    556691     23005   printing smudges/thrown away     -9600   
345179    567149    35607B                    thrown away        -9   
345180    567150    35817P                    thrown away       -44   
397212    571124    72802A         Wet pallet-thrown away      -306   
487975    577787     23055              rusty thrown away      -398   
487976    577788     23056               rusty throw away      -210   
487977    577789     23057               rusty throw away      -280   

             InvoiceDate  UnitPrice CustomerID         Country  
50849   2011-01-10 10:36        0.0          0  United Kingdom  
82794   2011-02-04 16:06        0.0          0  United Kingdom  
82796   2011-02-04 16:07        0.0          0  United Kingdom  
82797   2011-02-04 16:07        0.0          0  United Kingdom  
82799   2011-02-04 16:08        0.0          0  United Kingdom  
89118   2011-02-14 09:44        0.0          0  United Kingdom  
115489  2011-03-09 14:52        0.0          0  United Kingdom  
115504  2011-03-09 15:07        0.0          0  United Kingdom  
115505  2011-03-09 15:08        0.0          0  United Kingdom  
115818  2011-03-09 17:25        0.0          0  United Kingdom  
131398  2011-03-23 17:27        0.0          0  United Kingdom  
131399  2011-03-23 17:28        0.0          0  United Kingdom  
225528  2011-06-14 10:36        0.0          0  United Kingdom  
225529  2011-06-14 10:37        0.0          0  United Kingdom  
225530  2011-06-14 10:37        0.0          0  United Kingdom  
345179  2011-09-16 15:29        0.0          0  United Kingdom  
345180  2011-09-16 15:29        0.0          0  United Kingdom  
397212  2011-10-14 09:36        0.0          0  United Kingdom  
487975  2011-11-21 16:28        0.0          0  United Kingdom  
487976  2011-11-21 16:28        0.0          0  United Kingdom  
487977  2011-11-21 16:28        0.0          0  United Kingdom

In [29]:
# Seems like 'thrown away' filter is unnecessary as all these records have UnitPrice equals 0

In [30]:
# Filtering by zero UnitPrice 
df[df.UnitPrice == 0]

InvoiceNo StockCode                   Description  Quantity  \
622       536414     22139                             0        56   
1970      536545     21134                             0         1   
1971      536546     22145                             0         1   
1972      536547     37509                             0         1   
1987      536549    85226A                             0         1   
...          ...       ...                           ...       ...   
536981    581234     72817                             0        27   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538554    581408     85175                             0        20   
538919    581422     23169                       smashed      -235   

             InvoiceDate  UnitPrice CustomerID         Country  
622     2010-12-01 11:52        0.0          0  United Kingdom  
1970    2010-12-01 14:32        0.0          0  United Kingdom  
1971    2010-12-01 14:33        0.0          0  United Kingdom  
1972    2010-12-01 14:33        0.0          0  United Kingdom  
1987    2010-12-01 14:34        0.0          0  United Kingdom  
...                  ...        ...        ...             ...  
536981  2011-12-08 10:33        0.0          0  United Kingdom  
538504  2011-12-08 13:58        0.0          0  United Kingdom  
538505  2011-12-08 13:58        0.0          0  United Kingdom  
538554  2011-12-08 14:06        0.0          0  United Kingdom  
538919  2011-12-08 15:24        0.0          0  United Kingdom  

[2510 rows x 8 columns]

In [31]:
# Checking negative UnitPrice values
df[df.UnitPrice < 0]

InvoiceNo StockCode      Description  Quantity       InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  2011-08-12 14:51   
299984   A563187         B  Adjust bad debt         1  2011-08-12 14:52   

        UnitPrice CustomerID         Country  
299983  -11062.06          0  United Kingdom  
299984  -11062.06          0  United Kingdom

In [32]:
# Cutting out negative and zero UnitPrice values
df = df[df.UnitPrice > 0]

In [33]:
df.describe()

Quantity      UnitPrice
count  534125.000000  534125.000000
mean        9.916892       4.695887
std        40.787842      95.079545
min     -9360.000000       0.001000
25%         1.000000       1.250000
50%         3.000000       2.100000
75%        10.000000       4.130000
max      4800.000000   38970.000000

In [34]:
# Further investigating negative Quantity
df.sort_values(by='Quantity')

InvoiceNo StockCode                          Description  Quantity  \
4287     C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
160145   C550456     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR     -3114   
160144   C550456     21175          GIN + TONIC DIET METAL SIGN     -2000   
160143   C550456    85123A   WHITE HANGING HEART T-LIGHT HOLDER     -1930   
224419   C556522     22920                    HERB MARKER BASIL     -1515   
...          ...       ...                                  ...       ...   
160546    550461     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
270885    560599     18007  ESSENTIAL BALM 3.5g TIN IN ENVELOPE      3186   
97432     544612     22053                EMPIRE DESIGN ROSETTE      3906   
206121    554868     22197                 SMALL POPCORN HOLDER      4300   
421632    573008     84077    WORLD WAR 2 GLIDERS ASSTD DESIGNS      4800   

             InvoiceDate  UnitPrice CustomerID         Country  
4287    2010-12-02 14:23       0.03      15838  United Kingdom  
160145  2011-04-18 13:08       2.10      15749  United Kingdom  
160144  2011-04-18 13:08       1.85      15749  United Kingdom  
160143  2011-04-18 13:08       2.55      15749  United Kingdom  
224419  2011-06-13 11:21       0.55      16938  United Kingdom  
...                  ...        ...        ...             ...  
160546  2011-04-18 13:20       2.10      15749  United Kingdom  
270885  2011-07-19 17:04       0.06      14609  United Kingdom  
97432   2011-02-22 10:43       0.82      18087  United Kingdom  
206121  2011-05-27 10:52       0.72      13135  United Kingdom  
421632  2011-10-27 12:26       0.21      12901  United Kingdom  

[534125 rows x 8 columns]

**Something wrong with these -9360 Rotation Silver Angels**

In [35]:
df[df.StockCode == '84347'].sort_values(by='UnitPrice')

InvoiceNo StockCode                          Description  Quantity  \
4287     C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
488440    577822     84347  ROTATING SILVER ANGELS T-LIGHT HLDR       600   
413848    572325     84347  ROTATING SILVER ANGELS T-LIGHT HLDR       600   
31595     538998     84347  ROTATING SILVER ANGELS T-LIGHT HLDR       480   
14773     537618     84347  ROTATING SILVER ANGELS T-LIGHT HLDR       120   
...          ...       ...                                  ...       ...   
30830     538881     84347  ROTATING SILVER ANGELS T-LIGHT HLDR        10   
21938     538148     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         5   
22046     538149     84347  ROTATING SILVER ANGELS T-LIGHT HLDR        10   
447050    574941     84347  ROTATING SILVER ANGELS T-LIGHT HLDR       236   
467817    576365     84347  ROTATING SILVER ANGELS T-LIGHT HLDR       135   

             InvoiceDate  UnitPrice CustomerID         Country  
4287    2010-12-02 14:23       0.03      15838  United Kingdom  
488440  2011-11-22 09:34       1.74      14607  United Kingdom  
413848  2011-10-24 09:54       1.74      14607  United Kingdom  
31595   2010-12-15 12:08       1.88      15061  United Kingdom  
14773   2010-12-07 13:50       1.88      15061  United Kingdom  
...                  ...        ...        ...             ...  
30830   2010-12-14 15:54       5.91          0  United Kingdom  
21938   2010-12-09 16:26       5.91          0  United Kingdom  
22046   2010-12-09 16:27       5.91          0  United Kingdom  
447050  2011-11-07 17:42       5.95          0  United Kingdom  
467817  2011-11-14 17:55       5.95          0  United Kingdom  

[468 rows x 8 columns]

In [36]:
angels = df[(df.StockCode == '84347') & (df.UnitPrice > 1)]
angels[angels['Quantity'] > 0].sum()['Quantity']

np.int64(9461)

In [ ]:
# This value is suspiciously close to |9360|... Was it a discount?

In [37]:
angels.Country.value_counts()

Country
United Kingdom    445
EIRE                8
Germany             4
Spain               2
Unspecified         1
Norway              1
Portugal            1
Czech Republic      1
Switzerland         1
Netherlands         1
Finland             1
France              1
Name: count, dtype: int64

In [38]:
angels[(angels['Quantity'] > 0) & (angels['Country'].isin(['United Kingdom']))].sum()['Quantity']

np.int64(9038)

In [39]:
# There's not enough evidence to prove my suspension, so I've decided to drop the row 
df = df.drop([4287])

In [40]:
df.describe()

Quantity      UnitPrice
count  534124.000000  534124.000000
mean        9.934435       4.695896
std        38.720511      95.079633
min     -3114.000000       0.001000
25%         1.000000       1.250000
50%         3.000000       2.100000
75%        10.000000       4.130000
max      4800.000000   38970.000000

In [41]:
df[df.UnitPrice == df.UnitPrice.max()]

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
222681   C556445         M      Manual        -1  2011-06-10 15:31    38970.0   

       CustomerID         Country  
222681      15098  United Kingdom

In [42]:
df[df.CustomerID == '15098']

InvoiceNo StockCode                     Description  Quantity  \
222670    556442     22502      PICNIC BASKET WICKER SMALL        60   
222680    556444     22502  PICNIC BASKET WICKER 60 PIECES        60   
222681   C556445         M                          Manual        -1   
222682    556446     22502  PICNIC BASKET WICKER 60 PIECES         1   
222692   C556448     22502      PICNIC BASKET WICKER SMALL       -60   

             InvoiceDate  UnitPrice CustomerID         Country  
222670  2011-06-10 15:22       4.95      15098  United Kingdom  
222680  2011-06-10 15:28     649.50      15098  United Kingdom  
222681  2011-06-10 15:31   38970.00      15098  United Kingdom  
222682  2011-06-10 15:33     649.50      15098  United Kingdom  
222692  2011-06-10 15:39       4.95      15098  United Kingdom

In [ ]:
print(38970.00/60) # =645.50 as suspected

Ok, the problem: Picnic Basket, 1 small piece costs 4.95 and has StockCode #22502
Picnic Basket, pack of 60 pieces costs 649.50 and had the same StockCode #22502
A UK's customer with ID #15098 buys 60 pieces of small basket, after 6 minutes finds another option of the same basket that comes as a pack of 60, which they like better. Buy the pack of 60 baskets, but makes a mistake a buy 60 packs, instead of one. Store manager helps correct the mistake (hence Manual input), reverse the transation, makes a correct one, and reverse the original option of sindgle baskets too.

Only the InvoiceNo = 556446 (row index #222682) is correct. The rest 4 rows can be savely deleted.

In [43]:
df = df.drop([222670, 222680, 222681, 222692])

**C stands for Corrected**

In [44]:
# Looks like "C" in InvoiceNo stands for "Corrected"
df[df['InvoiceNo'].str.startswith('C')].sort_values(by='UnitPrice')

InvoiceNo  StockCode                   Description  Quantity  \
108088   C545478          D                      Discount      -720   
182729   C552569          D                      Discount      -240   
355050   C567903          M                        Manual      -120   
540078   C581462      16219  HOUSE SHAPE PENCIL SHARPENER       -48   
287974   C562124      15034   PAPER POCKET TRAVELING FAN        -23   
...          ...        ...                           ...       ...   
16356    C537651  AMAZONFEE                    AMAZON FEE        -1   
15016    C537630  AMAZONFEE                    AMAZON FEE        -1   
43703    C540118  AMAZONFEE                    AMAZON FEE        -1   
43702    C540117  AMAZONFEE                    AMAZON FEE        -1   
524602   C580605  AMAZONFEE                    AMAZON FEE        -1   

             InvoiceDate  UnitPrice CustomerID         Country  
108088  2011-03-03 11:08       0.01      16422  United Kingdom  
182729  2011-05-10 12:06       0.03      12901  United Kingdom  
355050  2011-09-22 16:29       0.03      16422  United Kingdom  
540078  2011-12-08 18:51       0.06      12985  United Kingdom  
287974  2011-08-02 17:23       0.07      13081  United Kingdom  
...                  ...        ...        ...             ...  
16356   2010-12-07 15:49   13541.33          0  United Kingdom  
15016   2010-12-07 15:04   13541.33          0  United Kingdom  
43703   2011-01-05 09:57   16453.71          0  United Kingdom  
43702   2011-01-05 09:55   16888.02          0  United Kingdom  
524602  2011-12-05 11:36   17836.46          0  United Kingdom  

[9246 rows x 8 columns]

In [45]:
# all negative Quantities has actually marked with "C" in InvoiceNo
df[df['InvoiceNo'].str.startswith('C')] == df[df['Quantity'] < 0]

InvoiceNo  StockCode  Description  Quantity  InvoiceDate  UnitPrice  \
141          True       True         True      True         True       True   
154          True       True         True      True         True       True   
235          True       True         True      True         True       True   
236          True       True         True      True         True       True   
237          True       True         True      True         True       True   
...           ...        ...          ...       ...          ...        ...   
540449       True       True         True      True         True       True   
541541       True       True         True      True         True       True   
541715       True       True         True      True         True       True   
541716       True       True         True      True         True       True   
541717       True       True         True      True         True       True   

        CustomerID  Country  
141           True     True  
154           True     True  
235           True     True  
236           True     True  
237           True     True  
...            ...      ...  
540449        True     True  
541541        True     True  
541715        True     True  
541716        True     True  
541717        True     True  

[9246 rows x 8 columns]

In [46]:
# Rerunning this sells to filter out transactions related to store operation
#df[df['StockCode'].isin(['AMAZONFEE', 'S', 'POST', 'DOT', 'BANK CHARGES', 'PADS', 'B'])].sort_values(by='StockCode', ascending=False)

In [47]:
# Saving operational transactions into a separated subset
operation = df[df['StockCode'].isin(['AMAZONFEE', 'S', 'POST', 'DOT', 'BANK CHARGES', 'PADS', 'B'])]

In [50]:
# Removing operational transactions from the main dataset
df = df[~df['StockCode'].isin(['AMAZONFEE', 'S', 'POST', 'DOT', 'BANK CHARGES', 'PADS', 'B'])]
operation

InvoiceNo StockCode     Description  Quantity       InvoiceDate  \
45        536370      POST         POSTAGE         3  2010-12-01 08:45   
386       536403      POST         POSTAGE         1  2010-12-01 11:27   
1123      536527      POST         POSTAGE         1  2010-12-01 13:04   
1814      536544       DOT  DOTCOM POSTAGE         1  2010-12-01 14:32   
3041      536592       DOT  DOTCOM POSTAGE         1  2010-12-01 17:06   
...          ...       ...             ...       ...               ...   
541216    581494      POST         POSTAGE         2  2011-12-09 10:13   
541540    581498       DOT  DOTCOM POSTAGE         1  2011-12-09 10:26   
541730    581570      POST         POSTAGE         1  2011-12-09 11:59   
541767    581574      POST         POSTAGE         2  2011-12-09 12:09   
541768    581578      POST         POSTAGE         3  2011-12-09 12:16   

        UnitPrice CustomerID         Country  
45          18.00      12583          France  
386         15.00      12791     Netherlands  
1123        18.00      12662         Germany  
1814       569.77          0  United Kingdom  
3041       607.49          0  United Kingdom  
...           ...        ...             ...  
541216      18.00      12518         Germany  
541540    1714.17          0  United Kingdom  
541730      18.00      12662         Germany  
541767      18.00      12526         Germany  
541768      18.00      12713         Germany  

[2096 rows x 8 columns]

In [51]:
df.describe()

Quantity      UnitPrice
count  532024.000000  532024.000000
mean        9.966750       3.656122
std        38.792963      26.510409
min     -3114.000000       0.010000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        10.000000       4.130000
max      4800.000000    6930.000000

In [52]:
df[(df.CustomerID == '17315') & (df.StockCode == '20979') & (df.UnitPrice == 1.25)]

InvoiceNo StockCode                    Description  Quantity  \
433318    573911     20979  36 PENCILS TUBE RED RETROSPOT         6   
541717   C581569     20979  36 PENCILS TUBE RED RETROSPOT        -5   

             InvoiceDate  UnitPrice CustomerID         Country  
433318  2011-11-01 15:40       1.25      17315  United Kingdom  
541717  2011-12-09 11:58       1.25      17315  United Kingdom

### Discounts
StockCode = D, Description = Discount

In [53]:
# Saving discounts to a separate subset
discount = df[df.StockCode == 'D']

In [54]:
# Removing discounts from the original dataset
df = df[~(df.StockCode == 'D')]

### Returned Goods

___
### Investigating negative quantities
The hypothesis that returned products are recorded in the database with a negative `Quantity` values has proved itself with examples of records pairs (540421, 540422) and (61619, 61624).
Therefore, I cannot just delete all rows with `Quantity < 0`. I have to delete relevant positive transactions as well.

But first, I'd like to clean some more obvious rubbish.
___

Cleaning the data that irrelevant for the purpose 
not because it's negative but because it doesn't surve a purpose

In [55]:
# Separating mannual inputs, because it would need further investigation
manual = df[df.StockCode == 'M']

In [56]:
df = df[df.StockCode != 'M']

**Now only starting to work with this cycle**

In [57]:
df = df.sort_values(by='InvoiceDate', ascending=False)

In [58]:
df_neg = df[df.Quantity <0]

In [59]:
subset = df_neg[['StockCode', 'UnitPrice', 'CustomerID']]
neg_tuples = [tuple(x) for x in subset.to_numpy()]

In [60]:
df_pos = df[df.Quantity > 0]

In [61]:
subset_2 = df_pos[['StockCode', 'UnitPrice', 'CustomerID']]
pos_tuples = [tuple(x) for x in subset_2.to_numpy()]

In [62]:
matched = set(neg_tuples) & set(pos_tuples)

In [63]:
#df[df['StockCode', 'UnitPrice', 'CustomerID'].isin(matched)]
df_matched = df[pd.Series(list(zip(df.StockCode, df.UnitPrice, df.CustomerID)), index=df.index).isin(matched)]

In [64]:
df_rest = df.drop(df_matched.index)

In [70]:
index_to_del = [] #list of records to delete
for i in range(df_matched.shape[0]-1):
    if df_matched.iloc[i, 3] < 0:
        for j in range(i+1, df_matched.shape[0]-1):
            if (df_matched.iloc[i, 6] == df_matched.iloc[j, 6]) and (df_matched.iloc[i, 1] == df_matched.iloc[j, 1]) and (df_matched.iloc[i, 5] == df_matched.iloc[j, 5]) and (df_matched.iloc[j, 3]>0):
                df_matched.iloc[j, 3] += df_matched.iloc[i, 3]
                index_to_del.append(df_matched.index[i])
                break

In [71]:
df_matched[df_matched.Quantity > 0]

InvoiceNo StockCode                         Description  Quantity  \
541871    581585     22178     VICTORIAN GLASS HANGING T-LIGHT        12   
541876    581585     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        24   
541829    581579     22083           PAPER CHAIN KIT RETROSPOT         6   
541744    581571     21258          VICTORIAN SEWING BOX LARGE         8   
541745    581571     23168        CLASSIC CAFE SUGAR DISPENSER        36   
...          ...       ...                                 ...       ...   
60        536373    82494L         WOODEN FRAME ANTIQUE WHITE          6   
47        536372     22632           HAND WARMER RED POLKA DOT         6   
14        536367     84969  BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
8         536366     22632           HAND WARMER RED POLKA DOT         6   
6         536365     21730   GLASS STAR FROSTED T-LIGHT HOLDER         6   

             InvoiceDate  UnitPrice CustomerID         Country  
541871  2011-12-09 12:31       1.95      15804  United Kingdom  
541876  2011-12-09 12:31       0.85      15804  United Kingdom  
541829  2011-12-09 12:19       2.95      17581  United Kingdom  
541744  2011-12-09 12:00      10.95      15311  United Kingdom  
541745  2011-12-09 12:00       1.04      15311  United Kingdom  
...                  ...        ...        ...             ...  
60      2010-12-01 09:02       2.55      17850  United Kingdom  
47      2010-12-01 09:01       1.85      17850  United Kingdom  
14      2010-12-01 08:34       4.25      13047  United Kingdom  
8       2010-12-01 08:28       1.85      17850  United Kingdom  
6       2010-12-01 08:26       4.25      17850  United Kingdom  

[13892 rows x 8 columns]

In [72]:
df = pd.concat([df_rest, df_matched], ignore_index=False)

In [74]:
# These 1000 records of cancelled transactions don't have matched purchase records, so I'll count them as mistakes and delete them
df[df.Quantity < 0]

InvoiceNo StockCode                      Description  Quantity  \
540176   C581470     23084               RABBIT NIGHT LIGHT        -4   
540141   C581468     21314    SMALL GLASS HEART TRINKET POT       -10   
537601   C581323     21533         RETROSPOT LARGE MILK JUG        -1   
537496   C581305     22627              MINT KITCHEN SCALES        -1   
534310   C581145     23249    VINTAGE RED ENAMEL TRIM PLATE        -9   
...          ...       ...                              ...       ...   
1981     C536548     22077           6 RIBBONS RUSTIC CHARM        -6   
1442     C536543     22355        CHARLOTTE BAG SUKI DESIGN        -2   
939      C536506     22960         JAM MAKING SET WITH JARS        -6   
154      C536383    35004C  SET OF 3 COLOURED  FLYING DUCKS        -1   
99        536378     84991      60 TEATIME FAIRY CAKE CASES        -3   

             InvoiceDate  UnitPrice CustomerID          Country  
540176  2011-12-08 19:28       2.08      17924   United Kingdom  
540141  2011-12-08 19:26       2.10      13599   United Kingdom  
537601  2011-12-08 11:53       4.25      14442  Channel Islands  
537496  2011-12-08 11:42       8.50      16933   United Kingdom  
534310  2011-12-07 13:48       1.49      17490   United Kingdom  
...                  ...        ...        ...              ...  
1981    2010-12-01 14:33       1.65      12472          Germany  
1442    2010-12-01 14:30       0.85      17841   United Kingdom  
939     2010-12-01 12:38       4.25      17897   United Kingdom  
154     2010-12-01 09:49       4.65      15311   United Kingdom  
99      2010-12-01 09:37       0.55      14688   United Kingdom  

[8899 rows x 8 columns]

In [75]:
df = df[df.Quantity > 0]

In [76]:
df.describe()

Quantity      UnitPrice
count  519862.000000  519862.000000
mean       10.219945       3.275752
std        36.934743       4.277941
min         1.000000       0.040000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        11.000000       4.130000
max      4800.000000     649.500000

# Final Result

In [78]:
# Saving a clean dataset to a file
df.to_csv('online_retail_clean.csv', index=False)